First draft for augmented assignment expressions

ncoghlan opened the pull request.

ncoghlan committed:
First draft for augmented assignment expressions

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+        yield from (total := operator.iadd(total, value) for value in data)
+        return total
+
+The proposal in this PEP would allow the above code to work, while also allowing
```
What does "above" mean here?  The first or second example?  Given the verbosity of the second I'd hope it's at least the first. :)

ncoghlan replied:
It means they'd both work :)

ericsnowcurrently replied:
That's what I figured (and what becomes more clear by the time one finishes reading the PEP). :)  That said, changing "the above code" to something like the more specific "the two preceding examples" would help avoid any confusion.

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+    while remaining_attempts -= 1:
+        ... # Attempt the operation once
+
+With a preceding local name binding or declaration, it would allow name binding
```
FWIW, the condition here did not jump out at me until the second example where you explicitly "declared" a local variable.  It may be worth putting this statement into a clearer direct statement, since it's a key part of the proposal.

ncoghlan replied:
As in, you think the requirement for a preceding local name binding or declaration needs to be clearer?

ericsnowcurrently replied:
Sort of.  In particular, the idea of an explicit declaration (i.e. using an annotation) was not obvious yet at this point in the text.  However, looking at it again I'm not sure that's the issue.

I suppose the actual problem I had is in getting distracted from the example.  Something like the following would help keep the focus on the enabled use case and example (as well as provide a strong visual contrast with the status quo):

    Currently if/else chains can lead to heavy nesting::

        m = pattern.match(data)
        if m:
            prep1()
        else:
            m = other_pattern.match(data)
            if m:
                prep2()
            ....
        use_match(m)

    Under this proposal the code would become cleaner::

        m = pattern.match(data)
        if m:
            prep1()
        elif m := other_pattern.match(data):
            prep2()
        elif ...
        use_match(m)

    If the name was already bound (or declared) then the first two lines can be combined::

        m : typing.Match  # a "declaration" for a local variable
        ...
        if m := pattern.match(data):
            prep1()
        ...

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+        ...
+
+    # Or equivalently
+    m : typing.Match
```
Whoa, I thought this syntax was only supported in class definitions.  I guess I didn't follow that change very closely. :)

ncoghlan replied:
I double checked in PEP 526 - annotations are allowed anywhere, but they'll only be stored in `__annotations__` at module and class scope.

ericsnowcurrently replied:
Oh, I double-checked too. :)  You're fine.  It's actually a good application of the syntax.  I just hadn't realized this was supported outside class definitions.  TBH, it weirds me out a bit.  (/me shudders thinking about variable declarations everywhere.) :)

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+Note that this PEP also delegates the task of deciding whether or not it's a
+good idea to implicitly allow nonlocal name rebinding operations to PEP 572 -
+it's a competing PEP aimed at ensuring the resulting capability is at least
+moderately teachable (think "If we do this at all, we should do it this
```
This parenthetical comment is a bit unclear and distracting.  In fact, this entire paragraph is a little raw and unfocused.  Still, the point you're making here is an important one.  My initial impression, prior to this paragraph, was that this PEP was meant to obviate 572.  Regardless, consider giving the relationship with 572 its own section or subsection.  That will help distinguish the topic and drive its focus a little better.

ncoghlan replied:
Yeah, that's fair - the snark doesn't help clarify anything.

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+--------------------------------
+
+The language grammar would be adjusted to allow augmented assignments to appear
+as expressions, where the result of the expression is the same reference that
```
Ah, so not just single identifiers.  Given the amount of attention that PEP 472 (and ensuing related discussions) has received, it may be worth making this particular detail more prominent.  Concise (1-2 lines) examples of each kind of binding (e.g. identifier, attribute, key, index) would help.

....ah, you do keep the identifier-only restriction in the expression case.  It may be worth making that more clear here then.

ncoghlan replied:
The restriction is in the very next paragraph, and both paragraphs are only single sentences, so I'm not sure how much more prominent I can make it...

ericsnowcurrently replied:
I guess the word "reference" threw me off; that and my previous awareness of restricting assignment expressions to identifiers (i.e. from the PEP 472 discusssions).  I'm a very impatient reader sometimes.  I'm sure you're fine here. :)

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+   means (i.e. a function parameter declaration, a variable declaration, a
+   regular assignment statement, or another name binding operation), then the
+   local scope is used.
+4. If implicitly declaring the augmented assignment target as ``nonlocal`` would
```
This is the key scoping change (aside from AmbiguousTargetError), right?  I am unsure of the utility of the status quo and consequently whether or not this change would lead to problems in existing code.  It certainly deviates from my normal expectations for scope resolution relative to augmented assignment.  However, I think I see where you're heading with this and plain assignment expressions. :)

Anyway, on the one hand, (unlike normal assignment) the augmented form implies a pre-existing name, so the proposed implicit "nonlocal" is relatively intuitive.  On the other hand, this has the possibility of masking a genuine NameError where the programmer (unfortunately) has bound the same name in an outer scope.  The likelihood of that is relatively low, but there is precedent for conceding on this concern in other scope resolution cases.

ncoghlan replied:
Yep, and if it bothers Guido too much, then we can either restrict doing this implicitly to scopes that are themselves implicit (i.e. comprehensions and generator expressions), or else restrict it to one level of implicit scoping, and require explicit scoping beyond that (which would still be enough to allow `add_one = lambda: tally += 1`).

Either of those two variants have the virtue of playing nice with global scopes, as well as preventing implicitly "jumping over" class scopes (you'd have to explicitly jump over them by declaring a name as explicitly non-local in a method body).

ericsnowcurrently replied:
Any extra complexity in the scoping rules beyond what you've proposed is going to make this a harder sell.  So I'd be careful there.  It might be good enough to add something like a ``__future__.implicit_nonlocal`` for just this case, to bring an extra measure of awareness and to give folks a chance to check for (unlikely) backward incompatibility.

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+        return total
+
+To address this, the compilation process will be changed to handle augmented
+assignment target scope resolution as follows:
```
After reading the rules, I found myself wishing I had examples to reference for each.  Mentioning the later examples here would help.  So would having at least a short example of each.

ncoghlan replied:
Yeah, that's a good idea.

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+the implicit ones added for augmented assignment targets) will be changed to take
+explicit ``global NAME`` declarations into account, such that the affected name
+is considered ``global`` in the inner scope as well. For example, the
+following code would work by binding ``x`` in the global scope instead of
```
+1

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+
+The issue is that name resolution for module level code is handled independently
+for each statement, so even though this is clear to a *human*, it's ambiguous
+to the compiler. Binding a new module global (and hence exporting a new
```
good point

ericsnowcurrently reviewed (commented):
Overall I think you've provided a solid proposal.  I find the extended parallel here between plain and augmented assignment to be a bit more compelling and palatable than only-assignment-expressions. :)  I'd certainly use the syntax on occasion.

My main concern with the content is that some of the important details are not clear at first.  Several key points were not clear until one or two sections after you introduced them.  Perhaps the abstract could be re-worked to summarize these better, or consider adding a quick point-by-point summary section (right after the abstract).

Other than that, I've left comments in-line.

ncoghlan commented:
A few useful references from my mailing list scoping discussions with Tim:
* https://mail.python.org/pipermail/python-dev/2018-May/153564.html (about name bindings in synthetic scopes in general)
* https://mail.python.org/pipermail/python-ideas/2018-May/050846.html (responding to Tim's comments on my initial sketch of the idea in this PEP)
* https://mail.python.org/pipermail/python-ideas/2018-May/050729.html (language ref level explanation of his proposed semantics)

Next steps:
* update the PEP to explicitly define the notion of "inline scopes" (lambda expressions, generator expressions, comprehensions) and "containing scopes" (modules, classes, function definitions)
* state that augmented assignment expressions always affect their containing scope, not their inline scope, but will *not* implicitly define new local variables in that scope - instead, they will raise `AmbiguousTargetError`
* for now, containing class scopes are always judged to be ambiguous for this purpose (unless directly in the class body), and hence augmented expressions simply aren't allowed as part of inline scopes inside a class body
* for containing module scopes, the target is always global, and implicitly declared global in any inline scopes
* for containing function scopes, the target is required to be declared as local, nonlocal or global in that scope
    * both variable declarations (in the type annotation sense) and name bindings count as declaring a name as local to the function
    * unlike module and class scopes, this declaration requirement is enforced at compile time, even for augmented assignment expressions directly in the function body (however, ambiguous targets for existing augmented assignments will only raise DeprecationWarning)
    * this gives an explicit behavioural difference between `TARGET = EXPR` and `TARGET := EXPR`: while the former can be used anywhere to bind a new local variable, the latter will complain about the target being ambigous if you try to use it at function scope without otherwise establishing that the intended target is a local variable.

ericsnowcurrently commented on pep-9999.rst:
```diff
@@ -0,0 +1,247 @@
+With a preceding local name binding or declaration, it would allow name binding
+in if/elif chains::
+
+    m = pattern.match(data)
```
FWIW, my initial inclination (based on this example) is to combine these two lines as-is (without any preceding locals-binding):

    if m := pattern.match(data):
        ...

It may be worth adding a comment to the example to reassure readers that there's a reason the code is the way it is:

    # We explicitly bind locally to avoid ambiguity (see the scope resolution rules below).
    m = pattern.match(data)
    if m:
        ...

This isn't that big a deal, but I figured I'd point it out so you can weigh my experience against the cost of extra text. :)

ncoghlan commented:
Mike Miller's survey of more modern languages (which have all moved away from assignment-as-expression in favour of assignment-as-statement, with limited exceptions for conditionals and loops): https://mail.python.org/pipermail/python-ideas/2018-May/050920.html

ncoghlan committed:
Revising as per PR discussion

ncoghlan committed:
Restrict := to single names everywhere

ncoghlan committed:
Remove mostly irrelevant paragraph

ncoghlan commented:
I've updated the draft to match the semantics proposed above, and clarified various details in the process.

The examples section still needs to be rewritten, and the design discussion section on handling augmented assignments in scoped expressions needs to be fleshed out.

A couple of bikeshed colours also need feedback:

* I decided `TargetNameError` was a better exception, since it avoids the jargony word `Scope`, and allows for errors that aren't specifically related to ambiguity (like just flat out prohibiting some scoping combinations)
* I currently separate the scope types as "block scopes"/"containing block scopes" and "scoped expressions", mainly due to the fact that "in an inline scope" ends up becoming an unintentional tongue twister due to all the "n"s in close proximity to each other.

ncoghlan committed:
Miscellaneous updates

ncoghlan committed:
More wording adjustments

ncoghlan committed:
Typo fix

ncoghlan committed:
Getting ready for first posting to python-ideas

- fill in TODO section and add references
- redo the examples section
- general edits and updates

ncoghlan commented:
@ericsnowcurrently I think this is basically publication-ready now, so it would be handy if you could take another look.

No worries if you don't have time though - this has already had more pre-review than most of my PEPs have received prior to first posting :)

ncoghlan committed:
Typo fix

ncoghlan committed:
After is not before

ncoghlan committed:
Explain the technical reasons

ncoghlan commented:
I realised this morning that augmented assignments used as actual expressions inside a larger statement can also raise TargetNameError from the beginning - it's only when they're used as standalone statements (which is already legal syntax) that it needs to be DeprecationWarning for at least one release.

ericsnowcurrently reviewed (approved):
Looks great!  I've found this update to be substantially easier to follow that the first draft. :)  Nice job.

ncoghlan committed:
Final pre-publication wording tweaks

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into augmented-assignment-expressions.rst

ncoghlan committed:
Assign PEP 577 a number + markup fixes

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+For example, this will allow operation retry loops to be written as::
+
+    remaining_attempts = 10
+    while remaining_attempts -= 1:
```
Doesn't this have a bug in it? I think it only makes 9 attempts. Consider starting with `remaining_attempts = 1`, then the `-=` operator decrements it to zero, and the loop bails out immediately. Sounds like you need `while (remaining_attempts -= 1) >= 0:` instead.

ncoghlan replied:
Yeah, you're right - rather than giving up the nice attribute name, or making the condition more complex, I'll just add a `+ 1` with a comment in the initialiser.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+default to targeting the nearest containing block scope.
+
+A new compile time ``TargetNameError`` is added as a subclass of ``SyntaxError``
+to handle cases where it either isn't clear to the compiler which target is
```
I'd leave the compiler out of it. This should be fully specified, not left as an implementation detail.

ncoghlan replied:
It's actually fully specified in the rest of the PEP - I was just anthropomorphising the compiler a bit here, rather than being clear and saying that *we're* the ones deeming particular cases to be ambiguous.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+As part of this, a semantic split is proposed between the handling of augmented
+assignments in regular block scopes (modules, classes, and functions), and the
+handling of augmented assignments in scoped expressions (lambda expressions,
```
We should at least carefully consider whether to treat `lambda` the same as comprehensions.

ncoghlan replied:
Yeah - one thing I like about treating it the same as comprehensions is the way it makes certain kinds of callbacks easier to write.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+expected to be rebound by an augmented assignment, or else the augmented
+assignment target scope is invalid for another reason.
+
+Finally, ``NAME := EXPR`` is proposed as a name rebinding expression that
```
This phrasing sounds ominous. Does it really mean that this wouldn't work?
```py
def foo():
    (p := 1)
    print(p)
```
This would invalidate most examples from PEP 572 without a good reason. (The only one that would still work is the one about partial sums.)

ncoghlan replied:
It really means that that wouldn't work - you'd have to write it as either:

    def foo():
        p = 1 # Explicitly local
        print(p)

or: 

    def foo():
        p: typing.Any # Explicitly local
        p := 1 # Technically not a rebinding, which is why I'm considering just calling it a name binding operator
        print(p)

It's the same compile time restriction that other augmented assignment operators would have, and it's the restriction that keeps assignment expressions from encroaching too far onto the territory of assignment statements.

It also means that scoped expressions can't magically bring new local variables into existence: you *have* to have an assignment or declaration in the containing block scope, so there's no need to go spelunking through all the lambda expressions and comprehensions to find potential name binding operations.

(Now that your updates to PEP 572 have been merged, I do need to go through and look at the impact this restriction will have on their verbosity)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+The case for allowing inline assignments at all is made in PEP 572. This
+competing PEP was initially going to propose an alternate surface syntax
+(``EXPR given NAME = EXPR``), while retaining the expression semantics from
```
I don't know that this much historic detail is relevant to the readership of the PEP.

ncoghlan replied:
Yeah, this is mainly for folks that have been following along - I'll keep the section, but move it to the end and add a link from the abstract.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+As a result of that, and unlike PEP 572, this PEP focuses primarily on use
+cases for inline augmented assignment. It also has the effect of converting
+cases that currently inevitably raise ``UnboundLocalError`` at function call
```
I assume cases such as
```py
def foo():
    p += 1
```
right?

ncoghlan replied:
Yep, exactly.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+New syntax for a name rebinding expression (``NAME := TARGET``) is then added
+primarily as a lower level primitive to help illustrate, implement and explain
+the new augmented assignment semantics, rather than being the sole change being
+proposed.
```
This is weird though, since only plain assignment is needed in the majority of motivating examples, and you seem to treat it as the less useful variant here.

ncoghlan replied:
Yeah, I expect I'll reword this section after I've gone through the latest draft of PEP 572 and checked how the examples look with PEP 577 instead.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+The author of this PEP believes that this approach makes the value of the new
+flexibility in name rebinding clearer, while also mitigating many of the
+potential concerns raised with PEP 572 around explaining when to use
+``NAME = EXPR`` over ``NAME := EXPR`` (and vice-versa).
```
Please be explicit about the concerns you are mitigating here.

ncoghlan replied:
With this section moved to the end, folks reading in order will have seen the more detailed explanation of this in the Design Discussion section before they get to this reference (it's mainly the "Two ways to do it, with no clear criteria for choosing between them" objection, which PEP 577 resolves by saying "assignment expressions can't implicitly create a new name, they have to target an existing one", so that when you're defining a new name, you always want "=" or ":", while ":=" indicates that you expect the name to already be known in the current binding scope)

gvanrossum replied:
And the new PEP 572 resolves it by disallowing `:=` anywhere where `=` would be usable.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+augmented assignment expression is the same post-calculation reference as is
+being bound to the given target.
+
+For example::
```
So what's new here is that the result of `+=` is printed. That would become rather annoying if it were used for a large list or set though.

ncoghlan replied:
Yeah, I'll add an Open Question about that. Unlike `:=`, we're pretty heavily constrained as to the potential operand precedence of augmented assignment expressions (since the entire RHS has to be accepted as an expression in the statement form), so you couldn't even suppress the output just by appending `and ...` - you'd have to parenthesise the assignment as well.

gvanrossum replied:
Personally I think it's too cute to be a language feature. If it's cuteness people are after, they can parenthesize the assignment: `(a += 1)` *does* print the new value.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+Augmented assignments to attributes and container subscripts will continue to
+be restricted to the standalone statement form, and will be defined as
+returning ``None`` for purposes of interactive use. While a future PEP could
```
This sounds arbitrary and confusing. If `n += 1` prints the new value of `n`, then users will expect `obj.n += 1` to do the same.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+returning ``None`` for purposes of interactive use. While a future PEP could
+potentially make the case for allowing those more complex targets as expressions,
+this PEP doesn't attempt to do so due to the ambiguity around whether or not
+they should call ``__getitem__`` and/or ``__getattribute`` on the target
```
Typo: `__getattribute`.

Anyway, I'd resolve the ambiguity by always returning the result of the in-place operator (`__iadd__` etc.), since it always *does* return a result. IIRC there was already a discussion on what the result of `:=` should be in such cases and it was successfully argued (IIRC by analogy with `x = y = 0`) that it should always just return the value that was being assigned (regardless of what the object's `__setattr__` does).

ncoghlan replied:
I'd be fine with that - your suggestion here is what I had as a fallback plan in the Design Discussion section if this restriction was deemed arbitrary and confusing :)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+expression after performing the assignment, or else just return a reference to
+the binary operation result being assigned.
+
+(Note: as an implementation detail, the language grammar itself may allow all
```
That's not an implementation detail, it's an implementation strategy. The set of accepted programs should not be different regardless of how it's implemented.

ncoghlan replied:
The difference mainly shows up if you're just running a parser over code, without attempting to turn it into an AST. However, getting rid of the statement/expression distinction means this whole question can go away.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+or class scope (this also includes code executed using "exec" or "eval"). These
+will continue to implicitly declare a new local variable as the binding target
+as they do today, and (if necessary) will be able to resolve the name from an
+outer scope before binding it locally.
```
Ouch. This reveals ugliness in class and exec scopes that I had forgotten. :-( I'd love to see an exception that somehow doesn't make that worse, so that at some point in the future we could perhaps fix the ugliness.

ncoghlan replied:
If we added a new `LOAD_LOCAL` opcode that skipped the nonlocal/globals/builtin chain, then it might be possible to do something, but I'm not sure what a viable transition strategy would actually look like.

That said, I don't think either PEP 572 or 577 would make this any worse than it already is today.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+At function scope, augmented assignments will be changed to require that there
+be either a preceding name binding or variable declaration to explicitly
+establish the target name as being local to the function, or else an explicit
+``global`` or ``nonlocal`` declaration. ``TargetNameError``, a new
```
I assume the `nonlocal` declaration must be able to find a matching non-global definition in a containing function scope. Is that also the case for `global`? I.e. if I write
```py
# (No definition of p here)
def f():
    global p
    (p := 1)
f()
print(p)
```
Does that work or not? (It does work if we write `p = 1` instead of `(p := 1)`.)

ncoghlan replied:
That would work, for the same reason `(p := 1)` has to "just work" at global scope: the compiler has no real idea which names actually exist at global scope (especially when compiling in "single" mode). (Opt-in linters and type analysers could be more prescriptive on that front, though, and I actually suggest that they should be in the design discussion section)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+    x += 1 # Sets global "x" to 1
+
+    class C:
+        x += 1 # Sets local "x" to 2, leaves global "x" alone
```
How horrible that this works this way (not your fault). :-( Please at least add a comment stating that depending on this behavior is not recommended.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+        n += 1 # Raises UnboundLocalError at runtime
+
+    def local_declaration_without_initial_assignment():
+        n : typing.Any
```
Please use the recommended formatting without a space before the colon: `n: typing.Any`.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+potential problems in cases where an unused function would currently raise
+``UnboundLocalError`` if it was ever actually called, but the code is actually
+unused - converting that latent runtime defect to a compile time error qualifies
+as a backwards incompatible change that requires a deprecation period.
```
This is a good point about moving checks from runtime to compile time! (Same for unreachable blocks actually.)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+When augmented assignments are used as expressions in function scope (rather
+than as standalone statements), there aren't any backwards compatibility
+concerns, so the compile time name binding checks would be enforced immediately
+in Python 3.9.
```
Probably you meant 3.8 here?

It's important to address the syntactic ambiguity though -- a toplevel augmented assignment would be both an expression and a statement. (Possibly you already did and I've simply forgotten I'd already read your solution -- I was interrupted by much of a workday during this review.)

ncoghlan replied:
I was being overly casual with terminology and saying "statement" when I really meant "top level expression in an expression statement". I'll clarify that.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+introduce a new nested scope of execution, either as an intrinsic part of their
+operation (lambda expressions, generator expressions), or else as a way of
+hiding name binding operations from the containing scope (container
+comprehensions).
```
In the latest version of PEP 572 I introduced the term "comprehensions" to refer to container comprehensions as well as generator expressions. I like your "scoped expressions" for comprehensions + lambdas.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+hiding name binding operations from the containing scope (container
+comprehensions).
+
+Unlike regular functions, these scoped expressions can't include explicit
```
The "functions" here feels jarring, since in the previous section you conspicuously avoid using that word -- "scoped expressions" seems to distinguish between two forms of expressions, and functions are neither. Maybe the solution is to change "functions" to "function bodies"?

ncoghlan replied:
Yeah, I agree "function bodies" would be clearer - I'll change it.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+  scoped expression, will raise ``TargetNameError`` (this is a restriction that
+  could potentially be lifted later, but is being proposed for now to simplify
+  the initial set of compile time and runtime semantics that needs to be
+  covered in the language reference and handled by the compiler and interpreter)
```
Ah, good, we ended up making this a static error in PEP 572 too.

ncoghlan replied:
I'm pretty sure it was originally Tim's idea, and it simplifies a few things nicely.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+  the initial set of compile time and runtime semantics that needs to be
+  covered in the language reference and handled by the compiler and interpreter)
+
+For example, the following code would work as shown::
```
These examples (like many in this PEP) are useful to explain the precise semantics, but far from compelling as motivation.

ncoghlan replied:
Aye, most of the current examples are aimed more at "How would it work in practice?", and less at "Why would we do it?". The "Why" for me personally is that I'd soured completely on PEP 572 as a new standalone expression type, especially after the reaction from teachers at the Education Summit to Tim's scoping proposal was "Heck, no", but everything fell into place for me once I framed the proposed semantics in terms of `+=` instead - once I understood how `+=` would work as an expression, then grasping the semantics of `:=` was easy. Without that scaffolding though, `:=` remained a weird bolt-on that didn't feel like it meshed properly with the rest of the language.

(It's unfortunate I didn't think of this framing until during the sprints, so I didn't get to ask any of the Education Summit attendees about it in person. Once I've rolled your feedback into the PEP, I may start contacting folks directly to see if I can get their input that way)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+        incr_x = lambda: x += 1 # Error due to missing target "x"
+
+    def late_target():
+        incr_x = lambda: x += 1 # Error due to "x" being declared after use
```
Perhaps this should be a runtime error, since `x` is in the right scope? There are likely situations where the static analysis isn't so easy, e.g.
```py
def f():
    if bool():
        x = 0
    f = lambda: x += 1
```

ncoghlan replied:
The same thought occurred to me, since the equivalent explicit use of `nonlocal` would give `UnboundLocalError` at runtime, rather than a compile time error.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+at function scope) will be changed to take explicit ``global NAME`` declarations
+into account, such that the affected name is considered ``global`` in the inner
+scope as well. For example, the following code would work by binding ``x`` in
+the global scope instead of raising ``SyntaxError`` as it does today::
```
Why? I can see this for *implicit* nonlocal but not for explicit.

ncoghlan replied:
I was thinking I needed it for consistency - I wanted to be able to explain the way `lambda: x += 1` works as always implying `global x` at module scope and `nonlocal x` at function scope.

Technically I don't *have* to define or implement it that way - I could define it as the scope declaration in the function scope getting copied down into the implicit nested scope, such that the expansion included either `nonlocal` or `global`, depending on how the name was defined in the parent scope. I don't really like that, though.

gvanrossum replied:
That's pretty much how we do it in PEP 572, except of course it also has the special case that if there is neither `nonlocal x` nor `global x` in the containing scope, it becomes implicitly local in the containing scope (and non-local in the comprehension).

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+    1
+
+
+Adding an inline assignment expression
```
I made it up to here. Maybe I'll review the rest tomorrow.

ncoghlan replied:
Thanks for your feedback so far! You've nudged me across the line for a couple of design changes I was considering anyway (most notably, dropping the restriction to single names due to the fact that augmented assignment already supports attributes and subscripts), and highlighted a few areas that need clarification.

gvanrossum reviewed (commented):
Partial review. We have a lot in common!

ncoghlan reviewed (commented):
Thanks for the review comments! I'll be rolling in updates to address them as part of https://github.com/python/peps/pull/661 (the required edits are extensive enough that I won't be getting to them tonight, though)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+    >>> x |= 42
+    <__main__.Target object at 0x7f608caa8048>
+    >>> x.value
+    42
```
That's not a compelling example for the workaround (you'd have to show it inside a lambda or comprehension where it would actually be useful and readable).

But people wouldn't use something like that anyway. They'd continue to mutate a list or dict instead.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+Rather than requiring such workarounds, this PEP instead proposes that
+PEP 572's "NAME := EXPR" syntax be adopted as a new inline assignment
+expression that uses the augmented assignment scoping rules described above.
```
I still find Tim's reason for wanting inline assignment much better. He shows real world code where it would make the expression of the *idea* of a certain algorithm clearer, by avoiding boring repetition.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+Unlike existing augmented assignment statements, inline assignment expressions
+would be restricted entirely to single name targets (even when used as a
+standalone statement).
```
But why? (Other than that Tim proposed that when he was only thinking of `:=`?)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+would be to retain the current semantics of augmented assignment statements
+and have the expression result be the reference bound to the target (i.e.
+``__getitem__`` and ``__getattribute__`` would *not* be called after the
+assignment had already taken place)
```
This is the most logical choice and the one that should be made.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+When discussing possible binding semantics for PEP 572's assignment expressions,
+Tim Peters made a plausible case [1_,2_,3_] for assignment expressions targeting
+the containing block scope, essentially ignoring any intervening scoped
+expressions.
```
(Though he never mentioned lambda IIRC, and he hadn't thought about comprehensions inside comprehensions until I mentioned it. :-)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+- it includes lambda expressions in the set of scopes that gets ignored for
+  target name binding purposes, making this transparency to assignments common
+  to all of the scoped expressions rather than being specific to comprehensions
+  and generator expressions
```
You make this an argument of consistency, but that only works if one buys into the reasoning that "scoped expressions" are an important concept, and that lambdas are more closely related to comprehensions than to functions defined with `def`. I'm not sure I buy that, and I'd have to see more real world examples where a lambda's style is cramped by the inability to assign to variables to its containing scope.

Basically the difference here is that PEP 572 would allow `:=` in a lambda to assign to locals of that lambda, while your proposal adds an implicit `nonlocal` (or `global`) to such targets. It's not at all clear to me which would be more useful and less confusing. And I don't think that to the typical user lambda has much in common with comprehensions (the syntactic forms are entirely different, and so are the use cases).

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+once again difficult to detect the scoping difference between the outermost
+iterable expressions in generator expressions and comprehensions (you have to
+mess about with either class scopes or attempting to rebind iteration Variables
+to detect it), so there's also no need to tinker with that.
```
If I understand this paragraph correctly, you have this in common with PEP 572 -- there, too, an assignment expression occurring in the outermost iterable of a comprehension targets the same scope as an assignment expression occurring elsewhere in the comprehension.

I find it hard to imagine a non-obfuscated example where there's a lambda anywhere in a comprehension, so I don't care whether those are treated the same or differently.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+``NAME := EXPR`` are entirely semantically equivalent at every scope. This
+makes the two forms hard to teach, since there's no inherent nudge towards
+choosing one over the other at the statement level, so you end up having to
+resort to "``NAME = EXPR`` is preferred because it's been around longer".
```
Actually the latest version of PEP 572 makes top-level `:=` syntactically invalid, so there's a very clear nudge -- use `:=` in expressions and `=` at the top level.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+``NAME = EXPR`` form when possible, while also enabling the use of the
+``NAME := EXPR`` as a kind of simple compile time assertion (i.e. explicitly
+indicating that the targeted name has already been bound or declared and hence
+should already be known to the compiler).
```
I know that a compiler could do that, but I don't see a compelling usability argument for doing it. (And the syntactic hack in PEP 572 provides a similar incentive without requiring an explicit declaration.)

ericsnowcurrently replied:
As a reader/reviewer ``m := value`` is a clear indication that the name is already bound somewhere (much like augmented assignment).  That's something I'd find useful*.  Under the status quo you either miss that fact or have to search through the code.   So I'd consider it a valuable signal.

* ...particularly in projects with large source files/functions (and I don't have the option of driving better conventions)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+    # Post-decrementing a counter
+    remaining_attempts = 9
+    while remaining_attempts:
```
There's a potential bug in code where somehow the value ends up being negative and the loop never stops. This is annoying, and I've developed the habit of always writing such things as `while something >= 0` (or similar variants, but comparing to zero feels right).

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+Each of the available options hides some aspect of the intended loop structure
+inside the loop body, whether that's the state modification, the exit condition,
+or both.
```
I'm not sure that it's worth presenting two forms of such a long example -- it "buries the lede" so to speak. I'd keep just the clearest of the two alternatives, and simplify the body to
```py
if attempt_operation():
    break
```
I'd also drop the `else` clause on the `while` since it's the same for all versions of the example.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+    elif m := yet_another_pattern.match(data):
+        ...
+    else:
+        ...
```
The need to write the `if` clause different from the `elif` clauses is annoying in your version though. I find myself fairly frequently rearranging the clauses of such a "poor man's match statement" and having to adjust the pattern of today's top-most check just to satisfy the arbitrary requirement on `:=` makes me unhappy.

ericsnowcurrently replied:
FWIW, I agree that the spelling in the example isn't great.  I'd probably spell it with a declaration instead (as the PEP does in the next example).

That said, I share Guido's concern about requiring explicit declarations for single-use variables.  Plain assignment expressions can be useful in one-off contexts, where you won't use the variable outside the statement (e.g. in comprehensions).  In those cases having to pre-declare the variable would be annoying noise (both to read and to write).

So on the one hand I find the parallel with augmented assignment the most appealing think about this PEP.  It makes it easy to understand.  On the other hand the weakest aspect is requiring explicit declaration in those single-use cases.  Solving the latter without special-case rules (which would diminish the proposal's greatest strength) or extra syntax will be challenging.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+binding it to a value (as shown above), or else by including an appropriate
+explicit type declaration::
+
+    m : typing.re.Match
```
PEP 8

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+intermediate values in comprehensions and generator expressions by
+exporting them to the containing block scope::
+
+    factor = 1
```
I'd prefer to write this as `factor: int` since the initial value is never used.

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+        n //= factor
+
+    def cumulative_sums(data, start=0)
+       total = start
```
IMO this example would gain clarify if there was no `start` parameter and we just initialized `total = 0`. (Even better, don't show the function scaffolding at all -- people know how to turn an example into a function, and the PEP is carefully written so that this code will work regardless of whether it lives in a function or at the global level.)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+Allowing lambda expressions to act more like re-usable code thunks
+------------------------------------------------------------------
+
+This PEP allows the closure-based counter example::
```
I misread this as "counterexample" and was momentarily confused. :-)

gvanrossum commented on pep-0577.rst:
```diff
@@ -0,0 +1,652 @@
+
+    def make_counter(start=0):
+        x = start
+        return (lambda step=1: x += step)
```
But this example is straight from the Lisp and function language textbooks. I've *never* had to write one of these. When you change it from maintaining a single integer to more complex state that would make it useful in real code, it's much less likely that it still works as a lambda.

Also the `()` are redundant and I don't think they add anything here.

gvanrossum reviewed (commented):
We have an awful lot in common!

ncoghlan commented:
This became PEP 577
